### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by pwl**
Most of the apps repeatedly implement the same integrity tests so I moved them here.  You can run both the generic integrity tests and specific tests for the individual module with `python -m fastgenomics.app` (stub name).  I assume that these tests are run from within the docker, which simplifies the code.  This is a rough idea and it likely needs lot's of refinement so I'd welcome any comments.

### CODE REVIEW

**Comment by pwl** [comment]
A quick run down of proposed changes:

- Added classes for app, data, files and other abstractions.
- Class `FGApp` (`app.py`) reads and interprets the `manifest.json`, stores default parameters and checks the presence of license files, in other words anything `/app` related.
- Class `FGData` (`data.py`) stores the paths of subdirectories of `/fastgenomics`, reads and interprets `input_file_mapping.json` and `parameters.json`
- Class `FGProcess` (`process.py`) marries the `FGApp` to `FGData` and checks if they are both compatible.  This is the high-level interface exposed to the user.
-  The validation is now built into the constructors. so to validate an app you can simply call `FGApp(app_dir)`, with `app_dir` defaulting to `/app`.  The default paths can still be changed via environmental variables.  Same for `FGData(data_root)` and `FGProcess(app_dir, data_root)` to see if the app is compatible with the sample_data.
- Ther are also helper classes like `Files`, `InputFile`, `OutputFile` and `Parameters` that do what the names suggest.  These can be easily extended to add more functionality & type checking if necessary. For example, the `InputFile` could be extended with methods to read the contents of the specific file as a data frame.
- The backwards compatibility is provided through `deprecated.py`, there is a single global `FGProcess` defined there that gets updated after calling `set_paths` equivalent from the old implementation.  I'd deprecate this in the long run if the new implementation works reasonably well.
- I removed some tests that were focusing on the internals of the previous implementation and replaced them with the tests of the constructors.

File: fastgenomics/app.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        return manifest
+
+    def check_files(self):
+        for f in self._mandatory_files:
+            absolute_path = self.app_dir / f
+            if not absolute_path.exists():
+                raise FileNotFoundError(f"{f} not found in the app directory!")
+
+
+def check_manifest(config: dict):
```
Comment (ChristianSauer) [inline fastgenomics/app.py]: I would probably first detect all errors and then raise an combined exception. Spares the user from fixing error 1, hitting run, fixing error 2..etc
But not mandatory, feel free to ignore

File: fastgenomics/app_creator.py [hunk 2]
```diff
@@ -11,14 +11,19 @@
 from . import _common
 
 
-logger = getLogger('fastgenomics.testing')
+logger = getLogger("fastgenomics.testing")
 
 
 # registry
-DOCKER_REGISTRY = 'apps.fastgenomics.org'
+DOCKER_REGISTRY = "apps.fastgenomics.org"
```
Comment (ChristianSauer) [inline fastgenomics/app_creator.py]: This registry does not exist.
Force the user to add a valid registry! E:g. leave this blank

File: fastgenomics/data.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+fastgenomics specification.  It also loads the input file mappings and
+checks if the files exist.
+
+    """
+
+    _subdirs = ["data", "config", "output", "summary"]
+
+    def __init__(self, data_root=DEFAULT_DATA_ROOT):
+        if isinstance(data_root, str):
+            data_root = Path(data_root)
```
Comment (ChristianSauer) [inline fastgenomics/data.py]: As above, check if it exists?

File: fastgenomics/data.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        mapping_file = self.paths["config"] / "input_file_mapping.json"
+        return json.loads(mapping_file.read_bytes())
+
+    def get_paths(self):
+        return {dir: self.root / dir for dir in self._subdirs}
+
+    def get_parameters(self):
+        params_file = self.paths["config"] / "parameters.json"
+        if params_file.exists():
+            return json.loads(params_file.read_bytes())
```
Comment (ChristianSauer) [inline fastgenomics/data.py]: read_bytes? Why not read_text?

File: fastgenomics/data.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        else:
+            return {}
```
Comment (ChristianSauer) [inline fastgenomics/data.py]: I think, that we previously returned the default parameters from the manifest

File: fastgenomics/external/anndata.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
+try:
+    import pandas as pd
+    import numpy as np
+    import scipy.sparse as sp
+    from anndata import AnnData
+except ImportError:
+    raise ImportError(
```
Comment (ChristianSauer) [inline fastgenomics/external/anndata.py]: What happens if anndata installed but not numpy?
Can happen on borked installs.... Better raise the existing error, too

File: fastgenomics/external/anndata.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        "This module requires anndata (https://github.com/theislab/anndata)."
+    )
+
+
+def get_path(file, content_type):
+    """Returns a path iff the file is of requested type and """
```
Comment (ChristianSauer) [inline fastgenomics/external/anndata.py]: Typo

Comment (ChristianSauer) [reply fastgenomics/external/anndata.py]: and??? :)

File: fastgenomics/io.py [hunk 8]
```diff
@@ -1,298 +1,107 @@
...
+                path=path,
+                optional=isoptional,
+            )
+
+        not_in_manifest = set(mapping) - set(specs)
+        if not_in_manifest:
+            logger.warning(
+                f"Ignoring files defined in input_file_mapping.json: {not_in_manifest}"
+            )
+
```
Comment (ChristianSauer) [inline fastgenomics/io.py]: I would probably add a logger.debug call and dump all files you have found, simplifies error detection

File: fastgenomics/io.py [hunk 9]
```diff
@@ -1,298 +1,107 @@
...
+                type=spec["type"],
+                usage=spec["usage"],
+                path=Path(root) / spec["file_name"],
+            )
+        self.files["summary"] = FileOutput(
+            name="summary", type="summary", usage="Summary file", path=summary_path
+        )
+
+    def clean(self):
+        """Deletes all the output files.  Useful for cleanup before tests."""
```
Comment (ChristianSauer) [inline fastgenomics/io.py]: Add a strong warning to not use it in production scenarios.
I dislike this method? Move to testing?

File: fastgenomics/io.py [hunk 10]
```diff
@@ -1,298 +1,107 @@
...
+    def __init__(self, specs, root, summary_path):
+        self.files = OrderedDict()
+        for name, spec in specs.items():
+            self.files[name] = FileOutput(
+                name=name,
+                type=spec["type"],
+                usage=spec["usage"],
+                path=Path(root) / spec["file_name"],
+            )
+        self.files["summary"] = FileOutput(
```
Comment (ChristianSauer) [inline fastgenomics/io.py]: add a const in defaults.py for summary?

File: fastgenomics/io.py [hunk 11]
```diff
@@ -1,298 +1,107 @@
...
-        name=param.name,
-        type=param.type,
-        value=new_value,
-        default=param.default,
-        optional=param.optional,
-        enum=param.enum,
-        description=param.description,
-    )
+    def __repr__(self):
+        return super().__repr__() + "IO type: Input" + "Optional: {self.optional}"
```
Comment (ChristianSauer) [inline fastgenomics/io.py]: Add newlines to keep the style of the original __repr__?
Also, missing f"" !

you could add an "IO Type" Field to the base class and populate it FileInput / Output. That way, you could omit the modified repr for FileOutput entirely and for FileInput, you would only add the Optional.

File: fastgenomics/parameters.py [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+            )
+        elif self.enum and val not in self.enum:
+            raise ValueError(
+                f"Setting parameter {self.name} with a value {val} that is not in {self.enum}."
+            )
+        self._value = val
+
+
+class FGParameters(object):
+    def __init__(self, parameters: dict):
```
Comment (ChristianSauer) [inline fastgenomics/parameters.py]: Acts like a dict...hmm add a keys() method?

File: fastgenomics/process.py [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+
+class FGProcess(object):
+    def __init__(self, app_dir=DEFAULT_APP_DIR, data_dir=DEFAULT_DATA_ROOT):
+        self.data = FGData(data_dir)
+        self.app = FGApp(app_dir)
+
+        self.parameters = self.app.default_parameters.copy()
+        self.parameters.update(self.data.parameters)
+
+        app_name = self.app.manifest["application"]["name"].lower().replace(" ", "_")
```
Comment (ChristianSauer) [inline fastgenomics/process.py]: Hmm, I would probably use a regex here and strip everything not in the A-z0-9. The name can by anything, which the logger probably does not like.
The replacing " " with "_" is totally valid though :)

File: fastgenomics/process.py [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+
+        self.input = FilesInput(
+            specs=self.app.inputs,
+            root=self.data.paths["data"],
+            mapping=self.data.input_file_mapping,
+        )
+        self.output = FilesOutput(
+            specs=self.app.outputs,
+            root=self.data.paths["output"],
+            summary_path=self.data.paths["summary"] / "summary.md",
```
Comment (ChristianSauer) [inline fastgenomics/process.py]: As above, use a default defined in defaults?

File: fastgenomics/app_creator.py (hunk 2 above)
Comment (pwl) [reply fastgenomics/app_creator.py]: That file was not used anywhere and contained obsolete functions anyway. Deleted.

File: fastgenomics/app.py (hunk 1 above)
Comment (pwl) [reply fastgenomics/app.py]: done and added tests.

File: fastgenomics/app.py [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+
+
+class FGApp(object):
+    # files that are checked for existence that are not otherwise
+    # explicitly used by the FGApp
+    _mandatory_files = ["LICENSE", "LICENSE-THIRD-PARTY"]
+
+    def __init__(self, app_dir=DEFAULT_APP_DIR):
+        if isinstance(app_dir, str):
+            app_dir = Path(app_dir)
```
[the remark this answers is not in the document]
Comment (pwl) [reply fastgenomics/app.py]: done

File: fastgenomics/data.py (hunk 5 above)
Comment (pwl) [reply fastgenomics/data.py]: by design, the `FGData` object has no access to the manifest.  The `{}` returned here is merged with the manifest defaults later on in `FGProcess`.

File: fastgenomics/data.py (hunk 4 above)
Comment (pwl) [reply fastgenomics/data.py]: my bad, fixed

File: fastgenomics/io.py (hunk 11 above)
Comment (pwl) [reply fastgenomics/io.py]: I fixe the new lines.

**Pull request merged**